### PR TITLE
Grahams Wu Tang Financial Shaolin Branch

### PR DESCRIPTION
**Pull request by grahamtn**

### CODE REVIEW

**Comment by grahamtn** [comment]
fixed scanner

File: src/main/java/io/zipcoder/wutangfinancial/Conversion.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package io.zipcoder.wutangfinancial;
+
+import static io.zipcoder.wutangfinancial.ConversionRates.*;
```
Comment (DavidGinzberg) [inline src/main/java/io/zipcoder/wutangfinancial/Conversion.java:3]: `import static` should be avoided in most cases (one exception would be the static import for JUnit's `Assert` functions. If you must call static methods from `ConversionRates` type the whole class name, particularly since `valueOf` is a method inherited from `Enum` and if you had more than one static import of enum classes you'd run into trouble.

File: src/main/java/io/zipcoder/wutangfinancial/Conversion.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+public class Conversion {
+
+
+    public Conversion(){
+
+    }
+
+    public double convert(double userAmount, ConversionRates moneyTypeGiven, ConversionRates moneyTypeExpected){
+        double givenRate = valueOf(String.valueOf(moneyTypeGiven)).getConversionRate();
```
Comment (DavidGinzberg) [inline src/main/java/io/zipcoder/wutangfinancial/Conversion.java:13]: No need to call `String.valueOf()` and then `ConversionRates.valueOf()`. You already had a `ConversionRates` object, just call `getConversionRate()` on it. (If you're trying to protect against null pointers, [this won't prevent them](https://docs.oracle.com/javase/7/docs/api/java/lang/Enum.html#valueOf(java.lang.Class,%20java.lang.String)). you'll need to check both objects aren't null beforehand)

File: src/main/java/io/zipcoder/wutangfinancial/Display.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        System.out.println(prompt);
+        int choice = sc.nextInt() - 1;
+        sc.nextLine();
+        return ConversionRates.values()[choice];
+    }
+
+    public static double getDouble(String prompt) {
+        do {
+            try {
+                sc = new Scanner(System.in);
```
Comment (DavidGinzberg) [inline src/main/java/io/zipcoder/wutangfinancial/Display.java:33]: You don't need a new `Scanner` object here. Using the one you already have should do just fine.

File: src/main/java/io/zipcoder/wutangfinancial/Display.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    public static ConversionRates getCurrencySelection(String prompt) {
+        System.out.println(prompt);
+        int choice = sc.nextInt() - 1;
+        sc.nextLine();
+        return ConversionRates.values()[choice];
+    }
+
+    public static double getDouble(String prompt) {
+        do {
```
Comment (DavidGinzberg) [inline src/main/java/io/zipcoder/wutangfinancial/Display.java:31]: `do/while` is valid, but not used very often and makes code slightly harder to read. Try to stick with ordinary `while` loops wherever practical

File: src/main/java/io/zipcoder/wutangfinancial/Display.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                Display.print("Please enter a number.");
+                continue;
+            }
+        } while (true);
+    }
+
+    public static void printCurrencies() {
+        int i = 0;
+        for (ConversionRates m : currencyGiven) {
+            String currencyName = ConversionRates.values()[i].getName();
```
Comment (DavidGinzberg) [inline src/main/java/io/zipcoder/wutangfinancial/Display.java:46]: You already have the value of `ConversionRates.values()[i]` in your `m` variable (which needs a better name). Use that instead of spending extra cpu time calling the `values()` function and then indexing the array again.

File: src/main/java/io/zipcoder/wutangfinancial/Display.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            System.out.printf("%d) " + currencyName + "%n", m.ordinal() + 1, m);
+            i++;
```
Comment (DavidGinzberg) [inline src/main/java/io/zipcoder/wutangfinancial/Display.java:48]: You almost never need a counter variable like this in a for each loop (aka "enhanced for loop"). try to avoid this and let the loop handle the iteration logic for you.

File: src/main/java/io/zipcoder/wutangfinancial/Display.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+                continue;
+            }
+        } while (true);
+    }
+
+    public static void printCurrencies() {
+        int i = 0;
+        for (ConversionRates m : currencyGiven) {
+            String currencyName = ConversionRates.values()[i].getName();
+            System.out.printf("%d) " + currencyName + "%n", m.ordinal() + 1, m);
```
Comment (DavidGinzberg) [inline src/main/java/io/zipcoder/wutangfinancial/Display.java:47]: Concatenation ( `aString + "foo" + otherString`) inside of a `printf()` or other formatting method is confusing and hard to follow. Pick one or the other and try not to mix them too much. You might try something like `printf("%d) %s%n", m.ordinal() + 1, currencyName, m)` to get the same output.

**Review by DavidGinzberg — COMMENTED** [review]
Your code is fairly clean; I gave some feedback where I saw potential improvements and tried to leave an explanation for each. Take a moment to make sure you also have a `.gitignore` file that ignores the `target/` directory, and learn how to remove that (and all the `*.class` files in it) from this repo -- that will help you a lot in the future.
